### PR TITLE
Allow Async actor to limit concurrency

### PR DESCRIPTION
**Pull request by siminm**
cc: @diranged

### CODE REVIEW

File: kingpin/actors/group.py [hunk 1]
```diff
@@ -328,6 +333,24 @@ class Async(BaseGroupActor):
...
+    def worker(self, act_gen, worker_id):
+        """Syncronous actor-executor that fetches from a shared generator.
+
+        Concurrency specifies the total number of workers. Each worker
+        continuously grabs the next task (actor) and executes it. Since the
+        acts come from the same pool (generator) no actions are executed twice.
+        """
+        for act in act_gen:
+            self.log.debug('Worker[%s]: %s' % (worker_id, act._desc))
+            yield act.execute()
```
Comment (siminm) [inline kingpin/actors/group.py]: Todo: figure out a fix for exception handling, such that the end-result is same as `yield [task, task, task]`

**Comment by siminm** [comment]
Reworked the logic to be simpler, and consistent with the rest of Tornado yield/raise error handling. No need to make this a separate module

File: kingpin/actors/group.py [hunk 2]
```diff
@@ -338,14 +349,18 @@ def _run_actions(self):
         failed (False).
         """
 
-        # This is an interesting tornado-ism. Here we generate and fire off
```
Comment (diranged) [inline kingpin/actors/group.py]: I wouldn't remove this comment -- its still very useful.

File: kingpin/actors/group.py [hunk 3]
```diff
@@ -338,14 +349,18 @@ def _run_actions(self):
...
         tasks = []
         for act in self._actions:
             tasks.append(act.execute())
 
+            if self.option('concurrency') > 0:
+                running_tasks = len([t for t in tasks if t.running()])
+
+                while running_tasks == self.option('concurrency'):
+                    self.log.debug('Concurrecnty saturated. Waiting.')
+                    yield gen.sleep(0)
```
Comment (diranged) [inline kingpin/actors/group.py]: I assume you want to sleep longer, right?

File: kingpin/actors/group.py [hunk 4]
```diff
@@ -338,14 +349,18 @@ def _run_actions(self):
...
-        # references to those tasks. However, we don't yield (wait) on them to
-        # finish.
         tasks = []
         for act in self._actions:
             tasks.append(act.execute())
 
+            if self.option('concurrency') > 0:
+                running_tasks = len([t for t in tasks if t.running()])
+
+                while running_tasks == self.option('concurrency'):
```
Comment (diranged) [inline kingpin/actors/group.py]: Just in case ... use `>=`?

File: kingpin/actors/group.py (hunk 2 above)
Comment (siminm) [inline kingpin/actors/group.py]: done

File: kingpin/actors/group.py (hunk 4 above)
Comment (siminm) [inline kingpin/actors/group.py]: done

File: kingpin/actors/group.py (hunk 3 above)
Comment (siminm) [inline kingpin/actors/group.py]: Ideally I'd have some `noop()` function. I don't actually need to sleep, I just need to allow the tornado stack to progress due to the internal ioloop. Basically making this equivalent to "do *something* then check again"

**Comment by siminm** [comment]
Addressed the comments and our IRL discussion. Ready for review again

**Comment by diranged** [comment]
:+1:

**Comment by diranged** [comment]
Would you consider adding this to the `rightscale.Execute` actor?

**Comment by siminm** [comment]
yay http://kingpin.readthedocs.org/en/latest/kingpin.actors.group.html?highlight=concurrency

**Pull request merged**